src/serializer/ResidualHeapVisitor.js

```diff
@@ -432,6 +439,47 @@ export class ResidualHeapVisitor {
     });
   }
 
+  // Addresses the case:
+  // let x = [];
+  // let y = [];
+  // function a() { x.push("hi"); }
+  // function b() { y.push("bye"); }
+  // function c() { return x.length + y.length; }
+  // Here we need to make sure that a, b both initialize x and y because x and y will be in the same
+  // captured scope because c captures both x and y.
+  _recordBindingVisitedAndRevisit(val: FunctionValue, residualFunctionBinding: ResidualFunctionBinding) {
+    let refScope = this.inAdditionalFunction ? this.inAdditionalFunction : "GLOBAL";
+    let funcToBindingState = getOrDefault(this.functionToBindingState, refScope, () => new Map());
+    let envRec = residualFunctionBinding.declarativeEnvironmentRecord;
+    invariant(envRec !== null);
+    let bindingState = getOrDefault(funcToBindingState, envRec, () => ({
+      capturedBindings: new Set(),
+      capturingFunctionsToCommonScope: new Map(),
+    }));
+    // If the binding is new for this bindingState, have all functions capturing bindings from that scope visit it
+    if (!bindingState.capturedBindings.has(residualFunctionBinding)) {
+      if (residualFunctionBinding.value) {
+        invariant(this);
+        for (let [functionValue, functionCommonScope] of bindingState.capturingFunctionsToCommonScope) {
+          invariant(this);
+          let prevCommonScope = this.commonScope;
+          this.commonScope = functionCommonScope;
+          let value = residualFunctionBinding.value;
+          this._withScope(functionValue, () => this.visitValue(value));
+          this.commonScope = prevCommonScope;
```

Comment:
This should probably be in a finally block.

Reply:
Definitely, I ran into an issue regarding this today when an error was thrown in `_withScope`.